Serhii jdbc

zaitsev-serhei opened the pull request.

zaitsev-serhei committed:
created base project structure to save the progress

zaitsev-serhei committed:
updated code to follow the requirements

updated ManufacturerDaoImpl to use try with resources

deleted unnecessary interface and its Impl

updated text for exceptions in ManufacturerDaoImpl

zaitsev-serhei committed:
updated code formatting following Maven tests

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+    private Connection connection = null;
+    //private PreparedStatement preparedStatement = null;
+    private ResultSet resultSet = null;
+    private DataBaseConnectionService dbConnector;
```
Эти переменные нам в классе не нужны, они могут быть локальные.

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        connection = dbConnector.getConnection();
```
connection тоже  открывай в TWR

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+            connection.close();
+        } catch (SQLException throwables) {
+            throw new DataProcessingException("Can`t create new Manufacturer !"
+                    + manufacturer.toString(), throwables);
```
```suggestion
        } catch (SQLException e) {
            throw new DataProcessingException("Can`t create new Manufacturer "
                    + manufacturer, e);
```

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+            throw new DataProcessingException("Can`t get Manufacturer with id "
+                    + id.toString(), throwables);
+        }
+        return Optional.of(manuf);
```
В случае ошибки вернешь Optional с new Manufacturer() внутри? Это не правильно

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+    public Optional<Manufacturer> get(Long id) {
+        Manufacturer manuf = new Manufacturer();
+        connection = dbConnector.getConnection();
+        String getStatement = "SELECT * FROM manufacturers WHERE id = ?";
```
![image](https://user-images.githubusercontent.com/20326858/235443428-a3352104-e077-4f5a-938e-60d19645d4b9.png)
Добавь условие в WHERE

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        String updateStatement = "UPDATE manufacturers SET name = ?, country = ?,"
+                + " isDeleted = ? WHERE id = ?";
```
isDeleted тут не нужен

Rommelua commented on src/main/java/mate/jdbc/dao/ManufacturerDao.java:
```diff
@@ -0,0 +1,19 @@
+
+    Manufacturer update(Manufacturer manufacturer);
+
+    boolean isDeleted(Long id);
```
![image](https://user-images.githubusercontent.com/20326858/235443585-19ee5621-857e-4af5-ad6f-964a5f128e9b.png)

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+        return manufacturer;
+    }
+
+    public ManufacturerDaoImpl() {
```
удали

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+
+    private Manufacturer getManufacturer(ResultSet resultSet) {
+        Manufacturer manufacturer = new Manufacturer();
+        try {
```
тут не нужен try catch

zaitsev-serhei replied:
при работе с резалт сет студия говорит обработать исключение

Sir-Beard replied:
> при работе с резалт сет студия говорит обработать исключение

мабуть мається на увазі, що треба переписати так, щоб студія не ругалась. 
Але загалом, думаю Льоня хоче, щоб ми навчились думати як архітектор

Rommelua replied:
Этот метод `getManufacturer `всегда вызывается внутри блока try catch. Соответственно в нем мы можем просто пробросить исключение выше, где оно и будет обработано)

Rommelua commented on src/main/java/mate/jdbc/model/Manufacturer.java:
```diff
@@ -0,0 +1,64 @@
+package mate.jdbc.model;
+
+public class Manufacturer {
+    private long id;
+    private String name;
+    private String country;
+    private boolean isDeleted;
```
В энтити классах всегда используем обертки, а не примитивы
```suggestion
    private Long id;
    private String name;
    private String country;
    private Boolean isDeleted;
```

Rommelua commented on src/main/java/mate/jdbc/model/Manufacturer.java:
```diff
@@ -0,0 +1,64 @@
+                + "id=" + id
+                + ", name='" + name + '\''
+                + ", country='" + country + '\''
+                + ", isDeleted=" + isDeleted + '}';
```
isDeleted тут вряд ли нужен

Rommelua commented on src/main/java/mate/jdbc/services/impl/DataBaseConnectionServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+
+@Dao
+public class DataBaseConnectionServiceImpl implements DataBaseConnectionService {
+    private Connection connection = null;
```
зачем эта переменная?

zaitsev-serhei committed:
updated code per requirements from first PR review

zaitsev-serhei committed:
updated SQL statement for Update query

Sir-Beard commented on src/main/resources/init_db.sql:
Use UPPERCASE for SQL queries.

Sir-Beard commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        String createStatement = "INSERT INTO manufacturers(name,country) VALUES (?,?)";
```
Use semicolon ";" at the end of sql statement**s**.
В майбутньому коли/якщо буде більше одного виклику до сервера одночасно, щоб не було ексепшенів.

Sir-Beard commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+            ResultSet resultSet = preparedStatement.getGeneratedKeys();
+            resultSet.next();
+            Long recordId = resultSet.getLong(1);
+            manufacturer.setId(recordId);
```
не бачу де в тебе закривається connection
сам собі відповідаю: воно не потрібно мабуть, через те що connection прописан в try-with-resources

Sir-Beard commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,137 @@
+    @Override
+    public Optional<Manufacturer> get(Long id) {
+        Manufacturer manuf = new Manufacturer();
+        String getStatement = "SELECT * FROM manufacturers WHERE id = ? AND isDeleted = false";
```
якщо в БД в колонці isDeleted буде null, то на цьому місці програма впаде
Тому, або додати щоб isDeleted була NOT NULL та мала дефолтне значення.
Або перевірку в методі що "має якесь значення".

zaitsev-serhei replied:
у БД установлено дефолтное значение

Sir-Beard commented on src/main/resources/init_db.sql:
```diff
@@ -0,0 +1,9 @@
+(
+    id        bigint auto_increment
+        primary key,
+    name      varchar(150)      null,
```
тут ти вказуєш що значення для name може бути null
в подальшому коли бд заповниться даними в тебе будуть дані з NULL'ами і це буде призводити до падіння програми при визові об'єкту.
тому, мабуть краще вказати not null, щоб людина точно заповнила поля якимись значеннями.

Sir-Beard commented on src/main/java/mate/jdbc/services/impl/DataBaseConnectionServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+            Connection connection = DriverManager.getConnection(DB_URL,USER_NAME,DB_PASSWORD);
+            return connection;
+        } catch (SQLException throwables) {
+            throw new DataProcessingException("Can`t connect to data base with credentials!"
```
постав space після знаку оклику, або переведи на наступну строку.
"Can`t connect to data base with credentials! "

zaitsev-serhei committed:
small changes to the code.
Updated SQL statements with semicolons
updated default value for columns in the DB

zaitsev-serhei committed:
refactored the code for Maven tests

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Override
+    public Optional<Manufacturer> get(Long id) {
+        Manufacturer manuf = new Manufacturer();
```
```suggestion
        Manufacturer manufacturer = new Manufacturer();
```

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+            ResultSet resultSet = preparedStatement.executeQuery();
+            if (resultSet.isBeforeFirst()) {
+                manuf = getManufacturer(resultSet);
+                return Optional.of(manuf);
+            } else {
+                return Optional.empty();
+            }
```
```suggestion
            }
            return Optional.ofNullable(manufacturer);
```

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+                 PreparedStatement preparedStatement = connection.prepareStatement(getStatement)) {
+            preparedStatement.setLong(1, id);
+            ResultSet resultSet = preparedStatement.executeQuery();
+            if (resultSet.isBeforeFirst()) {
```
```suggestion
            if (resultSet.next()) {
```

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+                 PreparedStatement preparedStatement =
+                         connection.prepareStatement(getAllStatement)) {
+            ResultSet resultSet = preparedStatement.executeQuery();
+            while (!resultSet.isLast()) {
```
```suggestion
            while (resultSet.next()) {
```

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+            while (!resultSet.isLast()) {
+                manufacturerList.add(getManufacturer(resultSet));
+            }
+            connection.close();
```
```suggestion
```

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+            preparedStatement.setString(2, manufacturer.getCountry());
+            preparedStatement.setLong(3, manufacturer.getId());
+            if (preparedStatement.executeUpdate() > 0) {
+                connection.close();
```
we use TWR)

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+            if (preparedStatement.executeUpdate() > 0) {
+                connection.close();
+                return manufacturer;
+            } else {
+                connection.close();
+                throw new DataProcessingException("Can`t update rows for Manufacturer"
+                        + manufacturer, null);
+            }
```
```suggestion
            }
                throw new DataProcessingException("Can`t update rows for Manufacturer"
                        + manufacturer, null);
```

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            preparedStatement.setLong(1, id);
+            if (preparedStatement.executeUpdate() > 0) {
+                connection.close();
```
same

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+        try (Connection connection = new DataBaseConnectionServiceImpl().getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            preparedStatement.setLong(1, id);
+            if (preparedStatement.executeUpdate() > 0) {
```
![image](https://github.com/Rommelua/jv-jdbc-ct/assets/20326858/2d8dbd19-69fd-4f60-970f-24d018fdef06)

Rommelua commented on src/main/java/mate/jdbc/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,140 @@
+        Manufacturer manufacturer = new Manufacturer();
+        try {
+            if (resultSet.next()) {
+                manufacturer.setId(resultSet.getLong("id"));
```
![image](https://github.com/Rommelua/jv-jdbc-ct/assets/20326858/f0de67ec-0c0d-496b-8f31-08da185d9649)

Rommelua commented on src/main/resources/init_db.sql:
```diff
@@ -0,0 +1,9 @@
+create table manufacturers
```
UPPERCASE

Rommelua commented on src/main/resources/init_db.sql:
```diff
@@ -0,0 +1,9 @@
+        primary key,
+    name      varchar(150)      not null,
+    country   varchar(50)       not null,
+    isDeleted tinyint default 0 null
```
```suggestion
    isDeleted tinyint default 0 not null
```

zaitsev-serhei committed:
refactored the code for Maven tests
updated SQL initial script
refactored DbConnectionUtil with new project structure and static method
refactored the code regarding PR instructions

zaitsev-serhei committed:
auto-formatting change of IntelliJ to TWR spacing ((((

Sir-Beard commented on src/main/java/mate/jdbc/dao/ManufacturerDao.java:
```diff
@@ -0,0 +1,19 @@
+import mate.jdbc.lib.Dao;
+import mate.jdbc.model.Manufacturer;
+
+@Dao
```
Не розумію навіщо нам анотація Dao на інтерфейсі.

zaitsev-serhei replied:
така анотація для того, щоб інжектор зміг знайти імліментаціїї інтерфейсів в файлах
